torch/csrc/jit/pickler.h

```diff
@@ -193,6 +194,31 @@ class Pickler {
     writer_(begin, sizeof(T));
   }
 
+  // Specialize 2 & 3 argument version of push as well, intended for small
+  // values. The motivation is that writer_ calls often incur per-call
+  // bounds/size checks. If we're just writing a few bytes, combine the sizes on
+  // the stack before sending off to the writer.
+  template <typename A, typename B>
```

Comment:
Would it be possible to instead define a version that takes an arbitrary number of arguments using a variadic template?

Reply:
Let me think about variadic templates idea - trying to think offhand how to handle the common buffer between the args cleanly. It's probably possible, but my question is whether the resulting code is going to be more readable (or more obfuscated) with the variadic templates idiom applied.

Reply:
I feel like a standard buffering approach to calling `write_` where we used a pickler-local 1kb buffer would get us the same or potentially more wins without requiring somewhat invasive API changes.

Reply:
Thanks for the feedback.

I don't think double-buffering necessarily makes sense here. The actual benchmark (and prod code) is just appending to a std::vector<char> behind the scenes. The main costs of doing this are effectively
  a) the branching related to bounds-checking and variable length memcpy.
  b) The std::function<> call overhead, effectively equivalent to a virtual function call

Double-buffering addresses (b) but not really (a).
The uploaded patch effectively addresses both issues, and it doesn't seem to be particularly complex.

fwiw, here's the resulting disassembly of the new pushInt, where we converted the 3 push<> calls to 1 for int64:
    P116779999